Ограничения на конверт культа и революции

TimErmolt opened the pull request:
- При успешном конверте мага в культ будут гибнуты все участвовавшие в конверте культисты. Маг при этом становится культистом и сохраняет свои способности. В начале процесса обращения всем участникам (магу и культистам) выводится сообщение, предупреждающее о последствиях в случае успешного конверта;
- Генокрадов, инопланетян (метроиды, нимфы дион и подобное), энтов дион, ксеноморфов и абдукторов нельзя конвертить в культ. Участвующим в конверте культистам выводится сообщение, что законвертить не получится;
- Остальные антаги при успешном конверте в культ теряют свои роли (и получают роль культа);
- Любых антагов нельзя приглашать в революционеры. Попытавшемуся пригласить выводится сообщение, что законвертить его не получится.

Протестить смог только конверт в культ от лица культиста. От лица жертвы не должно быть проблем, так как там только сообщения (культисты гибаются исправно, роли у антагов кроме мага и генокрада исчезают и выдаётся культист).
Конверт в революцию протестить не смог. Для определения наличия роли у цели там тот же код, что и в проверке на антага у культа, а там он работает, так что ужасных проблем быть не должно.

Сделано по иссую:
closes #1896

<details>
<summary>Чейнджлог</summary>

```yml
🆑TheUnknownOne
rscadd: Добавлены некоторые ограничения на обращение в культ и революцию:
tweak: За успешное обращение мага в культ культистам придётся пожертвовать собой (но маг не лишится своих способностей).
tweak: Генокрадов, дион, ксеноморфов, абдукторов и некоторых низших существ больше нельзя обращать в культ.
tweak: Прочие антагонисты при успешном обращении в культ лишатся своих прежних ролей.
tweak: Антагонистов (любых) теперь нельзя приглашать в революцию.
/🆑
```

</details>

- [x] Pull Request полностью завершен, мне не нужна помощь чтобы его закончить.
- [x] Я внимательно прочитал все свои изменения и багов в них не нашел.
- [ ] Я запускал сервер со своими изменениями локально и все протестировал.
- [x] Я ознакомился c [Guide to Contribute](https://github.com/ChaoticOnyx/OnyxBay/blob/dev/docs/contributing.md).

TimErmolt committed:
feat(cult, revolution): add convert restrictions

- Adds some restriction for converting people into cult and revolution.
- Cult: successfully converting a wizard (either forcefully with 2+ cultists or having them accept the offer) gibs the wizard;
- Cult: changelings, aliens, dionae, xenomorphs, and abductors can't be converted;
- Cult: other antags will lose their previous roles upon successful conversion.
- Revolution: antags can't be converted.

Note: This may mess with other antags that use the same conversion method as revolution, e.g. loyalists. Since we have them disabled, this shouldn't be an issue... for now.

TimErmolt committed:
refact(revolution): better convert restriction

- Implements the convert restriction in a better way, so that it doesn't mess with other antags/factions' converts.

TimErmolt commented:
Опять чужой код, который я вообще не трогал, рефакторить из-за чекера.

TimErmolt committed:
refact(cult runes): remove obsolete 'as'

- I have hate in my heart for DreamChecker.

TimErmolt commented:
Всё.

NefrNerf commented:
Кек, похоже на мой старый пр

NefrNerf commented:
Если что, вопрос гибага мага уже проработали и пришли к выводу, что лучше его конвертить и забирать роль, но по хорошему, никто его уже и не абузит.

TimErmolt commented:
> Кек, похоже на мой старый пр

Так работает кодеранск.

> Если что, вопрос гибага мага уже проработали и пришли к выводу, что лучше его конвертить и забирать роль, но по хорошему, никто его уже и не абузит.

А иссуй отредактировать не судьба. Впрочем, запилить легко, но я считаю, что гибать мага, попытавшегося заабузиться и стать культистом, либо обломать культистов, пытающихся провернуть ту же схему, -- справедливо и смешно. Особенно если это весьма редко будет происходить.
Пусть пока так остаётся, если г\*ймдиз\*йн будет однозначно против, то поменяю.

TobyThorne commented:
Смешнее и справедливее конвертить мага, но гибать при этом конвертящих культистов. Тогда у культа встаёт выбор между несколькими обычными культистами и культистом-визардом.

TimErmolt commented:
А ведь неплохой вариант. Если никто не возразит, завтра наверное сделаю.

Zert0X commented:
Хммм, где-то я это уже видел

Zert0X commented:
А зачем вообще было трогать реву если она технически отключена? Да и не рабочая в целом.
> Любых антагов нельзя приглашать в революционеры. Попытавшемуся пригласить выводится сообщение, что законвертить его не получится.

TimErmolt commented:
> А зачем вообще было трогать реву если она технически отключена? Да и не рабочая в целом.

Она вроде работает, просто в ротации режимов нет. И ходили слухи, что роль революционера иногда можно получить книжкой с астероида, и если это правда, то и революционеры в игре могут появиться и распространиться. Поэтому лучше сделать. Если нет, то можно отменить, хотя с тем же успехом это можно будет сделать в будущем, если примут решение это ограничение убрать. Не самая лучшая практика, конечно.
Короче, исходя из того, что по слухам на астероиде есть [книжка](https://github.com/ChaoticOnyx/OnyxBay/blob/efe5aea241d2cc688e75248136fcbdf8446f0c56/code/game/objects/items/contraband.dm#L60), с 50% шансом обращающая прочитавшего в революционера, пока оставлю.

TimErmolt committed:
tweak(cult): gib cultists instead of wizard

- Upon successfully converting a wizard, the participating cultists will get gibbed, while the wizard will be converted and maintain their wizard powers.

TimErmolt commented:
Сделяль.
- При успешном конверте мага в культ будут гибнуты все участвовавшие в конверте культисты. Маг при этом становится культистом и сохраняет свои способности.

DarkHunter7756 commented:
> * При успешном конверте мага в культ будут гибнуты все участвовавшие в конверте культисты. Маг при этом становится культистом и сохраняет свои способности.

![image](https://user-images.githubusercontent.com/76823962/161531717-baabdcac-1719-4c55-85f0-12cac98450ec.png)

А может им головы гибнуть, а не все тело? Будет атмосфернее

TimErmolt commented:
> А может им головы гибнуть, а не все тело? Будет атмосфернее

Сложно в коде сделать будет. Хотя скорее всего возможно.
На самом деле наверное так даже будет лучше, поскольку при полном гибе визарда будут забивать отлетающие кишки гибнутых культистов. Сегодня вечером попробую сделать.

DarkHunter7756 commented:
Есть еще проблема отображения гибнутой головы у пары сущностей по типу тонкотрапов. Да и выпадение мозга при гибе головы вроде не происходит.

TobyThorne commented:
> Есть еще проблема отображения гибнутой головы у пары сущностей по типу тонкотрапов.

Нету.

TimErmolt commented:
> Да и выпадение мозга при гибе головы вроде не происходит.

На самом деле выпадает. По крайней мере когда я гибнул голову трапу кулаками, он выпал.
Вообще стоит наверное сначала убивать мозг, а потом гибать голову, чтобы их не могли воскресить, успев засунуть мозг в ММИ.

DarkHunter7756 commented:
> На самом деле выпадает. 

ну у синтов не выпадает. Беру как пример )

TimErmolt committed:
tweak(cult runes): gib heads instead of bodies

- Only the cultists' heads will get gibbed, replacing the full-body gib. If a cultist is non-human, they will be gibbed regularly.

TimErmolt commented:
Таки сделал, оказалось даже легче чем я думал.
- Теперь культисты гибаются не полностью, а у них лишь сгорают головы. Нечеловеческие культисты гибаются как обычно.

NefrNerf commented on code/game/antagonist/antagonist_factions.dm:
```diff
@@ -12,6 +12,13 @@
+	var/datum/antagonist/antag
+	for(var/antag_type in GLOB.all_antag_types_)
+		antag = GLOB.all_antag_types_[antag_type]
+		if(antag.is_antagonist(M.mind))
```
А как же триторы? Они должны конвертится, сделай реализацию через новый флаг, что-то по типу CONVERTABLE

NefrNerf replied:
Чет 2 раза отправилось

NefrNerf commented:
А еще, замени 1 на TRUE, где это можно

TimErmolt commented:
Опять рефактор. Ладно, тогда уж целиком стиль кода приведу в порядок в этих двух файлах.

TimErmolt committed:
refact(cult runes, antag factions): spans and bools

- Changes HTML span tags to `SPAN()` macros, 0-1 to FALSE-TRUE, and some minor grammar fixes.

TimErmolt commented:
Вроде ничего не пропустил.

TimErmolt commented:
Блин, там что-то запилили, и из-за этого здесь теперь конфликты. Щас попробую починить.

TimErmolt committed:
Merge branch 'dev' into restricting-cult

TimErmolt committed:
fix(antag_factions): resolve conflicts fully

TimErmolt commented:
Всё, теперь должно быть нормально.

kreeperHLC commented:
Автор аж с апреля делает вещи и ещё жив. Вот это чудо.

Tsurupeta commented:
Снова конфликты

TimErmolt committed:
Merge branch 'dev' into restricting-cult

TimErmolt commented:
Всё. Можно пожалуйста уже два аппрува и мёрж, а то так и будет продолжаться же.
